🌐 Add WASM Support

zicklag opened the pull request:
Still, testing, but would you be open to supporting WASM? It looks like all that is needed is an extra dep for supporting `std::time::Instant` on web.

It's behind a feature flag so it doesn't effect anybody unconcerned with WASM.

zicklag commented:
Tested on my project and it looks like it works.

taiki-e commented on Cargo.toml:
```diff
@@ -11,8 +11,12 @@ documentation = "https://docs.rs/async-lock"
+
 [dependencies]
 event-listener = "2.5.1"
+instant = { version = "0.1.9", optional = true }
```
why does this not `target.'cfg(target_arch = "wasm32")'` dependencies?

zicklag replied:
🤦‍♂️ Yeah, that makes sense. Then we don't need the feature. Fixed! 👨‍🔧

zeenix commented:
LGTM!

taiki-e commented on Cargo.toml:
```diff
@@ -14,6 +14,9 @@ categories = ["asynchronous", "concurrency"]
 event-listener = "2.5.1"
 
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+instant = { version = "0.1.9" }
```
I think this has link issue like: https://github.com/Amanieu/parking_lot/issues/269

zicklag replied:
Hmm, then maybe we should just remove the use of `std::time::Instant` entirely when targeting WASM? That seems reasonable, because, like they said in the parking lot issue, WASM only has one thread so fair locking shouldn't be needed.

zicklag commented:
OK, I removed `instant` completely and just made sure that `Mutex` doesn't depend on `std::time` when targeting WASM. I didn't technically run tests on WASM, but if you take out the same code that is excluded when running on WASM then the tests pass on native.

I also added `cargo check` for WASM to CI!

taiki-e commented:
Thanks! I would like to run tests on CI if possible -- do you think it is possible to do something like https://github.com/smol-rs/fastrand/pull/7?

zicklag commented:
I'll take a look at it.

zicklag commented:
There you go! Tests are passing for WASM in CI.

taiki-e reviewed (approved):
LGTM. Thanks @zicklag!

zicklag commented:
No problem, glad I could help! I love the simple smol ecosystem. 😃

taiki-e commented:
Published in v2.4.0.
